### PR TITLE
Add "feedback" attribute to pl-multiple-choice

### PR DESCRIPTION
**Pull request by Cris-Moran**
Allows for developers to pass in a hint attribute to the `pl-answer` element in `pl-multiple-choice`. A visual demo of a hint is shown below:
Screen Shot 2021-08-10 at 3 51 00 PM
Potential issues:
- No way to display a hint next to answer choices created by `all-of-the-above` and `none-of-the-above` attributes using this  method, since they don’t appear as a pl-answer in question.html
- On some mobile devices, the hint and score badge can stick outside of their parent container.

### CODE REVIEW

**Comment by Cris-Moran** [comment]
**_I have reviewed and hereby sign the CLA_**

**Comment by mwest1066** [comment]
recheck

**Comment by mwest1066** [comment]
@Cris-Moran I'm sorry for the poor wording from the CLA bot, but it actually needs you to paste that text into a comment _without formatting_. Just the plain text "I have reviewed and hereby sign the CLA" (no quotes) in a comment. Thanks!

**Comment by Cris-Moran** [comment]
Ah okay, thank you for the clarification!

**Comment by Cris-Moran** [comment]
I have reviewed and hereby sign the CLA

File: elements/pl-multiple-choice/pl-multiple-choice.css
Comment (jonatanschroeder) [inline elements/pl-multiple-choice/pl-multiple-choice.css:12]: I wonder if it may be better to move all styling for this hint to the CSS (i.e., replace classes `mr-1 ml-1 pl-2` in the hint div with `margin-right`, `margin-left` and `padding-left` properties here). This could help with for consistency and make it easier to change in the future.

Also, I think it may be a good idea to introduce a light background or very thin border-top/bottom/right to distinguish it from the question itself.

File: elements/pl-multiple-choice/pl-multiple-choice.mustache
Comment (jonatanschroeder) [inline elements/pl-multiple-choice/pl-multiple-choice.mustache:18]:
```suggestion
        <div class="pl-multiple-choice-hint">
```

See comment in CSS file above. Also, this element doesn't need an `id`.

File: elements/pl-multiple-choice/pl-multiple-choice.py
Comment (jonatanschroeder) [inline elements/pl-multiple-choice/pl-multiple-choice.py:18]:
```suggestion
HINT_DEFAULT = None
```

File: elements/pl-multiple-choice/pl-multiple-choice.py
Comment (jonatanschroeder) [inline elements/pl-multiple-choice/pl-multiple-choice.py:228]:
```suggestion
                'display_hint': submitted_key == answer['key'] and answer['hint'] is not None,
```

**Review by jonatanschroeder — COMMENTED** [review]
This looks great! Can you please also delete the .DS_Store file from the commit?

Before this is approved, please:
* Describe the new attribute in the documentation (`docs/elements.md`);
* Add the attribute to some example course question (probably `exampleCourse/questions/element/multipleChoice/question.html`).

If you're feeling adventurous, would you be ok with adding a similar option in `pl-checkbox`? Maybe `pl-dropdown` too, let me know what you think.

File: elements/pl-multiple-choice/pl-multiple-choice.css
Comment (Cris-Moran) [reply elements/pl-multiple-choice/pl-multiple-choice.css:12]: Will do. I added a light gray background to the hint, and a border radius so the corners don't look sharp. I'll commit the changes shortly.
Screen Shot 2021-08-12 at 1 58 21 PM

**Comment by Cris-Moran** [comment]
I'll play around with adding hints for `pl-checkbox` and `pl-dropdown` as well. Once I figure out how to implement hints for them I'll submit more PRs.

**Comment by Cris-Moran** [comment]
Since the hint is now more distinguishable from the answer, I reduced the spacing between them by a little bit. Here's how it looks now:
Screen Shot 2021-08-13 at 11 27 12 AM

File: docs/elements.md
Comment (jonatanschroeder) [inline docs/elements.md:132]:
```suggestion
`hint` | string | — | Helper text to be displayed to the student next to the option after question is graded if this option has been selected by the student.
```

File: exampleCourse/questions/element/multipleChoice/question.html
Comment (jonatanschroeder) [inline exampleCourse/questions/element/multipleChoice/question.html:181]: You probably want to expand on what the hint is expected to convey, as a suggestion, and comment on what is expected to happen when an option with a hint is selected.

File: elements/pl-multiple-choice/pl-multiple-choice.py
Comment (jonatanschroeder) [inline elements/pl-multiple-choice/pl-multiple-choice.py:190]: Since you're changing this tuple to have 4 elements, you also need to update other places that handle this tuple, like lines 51 and 54, which get correct/incorrect answers from JSON files.

Comment (jonatanschroeder) [reply elements/pl-multiple-choice/pl-multiple-choice.py:190]: And lines 174 and 182, for all-of-these/none-of-these.

File: elements/pl-multiple-choice/pl-multiple-choice.css
Comment (jonatanschroeder) [inline elements/pl-multiple-choice/pl-multiple-choice.css:13]:
```suggestion
    background-color: rgba(0,0,0,0.05);
```
This is a suggestion, feel free to adjust if needed. The rationale for using opacity instead of color is that, if the background is set to another color (e.g., the instructor changed it to something else with CSS), this will still make the hint slightly darker than the background. A similar change may be a good consideration for the border as well.

**Comment by jonatanschroeder** [comment]
> * No way to display a hint next to answer choices created by `all-of-the-above` and `none-of-the-above` attributes using this  method, since they don’t appear as a pl-answer in question.html

This could be addressed with an element-level attribute, like `none-of-the-above-hint`, if you think this is important.

> * On some mobile devices, the hint and score badge can stick outside of their parent container.

I don't have an easy way to test this, can you include a screenshot? Though I don't think this would be a big problem.

File: elements/pl-multiple-choice/pl-multiple-choice.py
Comment (Cris-Moran) [reply elements/pl-multiple-choice/pl-multiple-choice.py:190]: Alright, I'll put `None` into the JSON file tuples in lines 51 and 54 for now since I'm not sure how hints will work with them at the moment, and I'll add something similar to the `none-of-the-above-hint` attribute you talked about which will be inserted into the tuple in lines 174 and 182.

**Comment by Cris-Moran** [comment]
> I don't have an easy way to test this, can you include a screenshot? Though I don't think this would be a big problem.

This issue only came up when I emulated a Galaxy Fold display in Chrome Dev Tools. As you can see the red box with the 'x' extends beyond its border:
Screen Shot 2021-08-13 at 5 43 12 PM

I don't have a Galaxy Fold to test this on. However, I tested the hints using Chrome on my own mobile device (which isn't a Galaxy Fold) and it looked exactly the same as when I selected it in the Chrome Dev Tools, so I believe it's accurate.

File: elements/pl-multiple-choice/pl-multiple-choice.css
Comment (Cris-Moran) [reply elements/pl-multiple-choice/pl-multiple-choice.css:13]: Good suggestion, I changed both the border and background to use this instead.

**Comment by jonatanschroeder** [comment]
We were talking about this feature today, and here are a few things that I'd like to ask you to change:

* Instead of "hint", call this "feedback". The idea of hint usually comes before submission, so to avoid confusion we'd like to clarify that this comes after submission.
* The `grade(data)` function in `pl-multiple-choice.py` should include this feedback as part of the `data['partial_scores']` object (together with score and weight). This would allow the feedback to be:
    * updated by the `grade` function of the question, which would allow instructors more flexibility in manipulating it;
    * better listed in logs;
    * included in downloaded files and updated via file upload.
* To ensure the update above can be used properly, the `render(data)` function in `pl-multiple-choice.py` should take the feedback from `partial_scores` (similar to the how the score for the badge is obtained) instead of from the answers list.
* This will probably come as consequence from the items below, but the feedback should only show after "Save & Grade", not after "Save Only".

**Comment by Cris-Moran** [comment]
Alright, just got done making the requested changes.

File: exampleCourse/questions/element/multipleChoice/question.html
Comment (jonatanschroeder) [inline exampleCourse/questions/element/multipleChoice/question.html:182]:
```suggestion
            <p>Here, you can observe the behavior of the <code>feedback</code> attribute in <code>pl-answer</code>. Feedback is expected to convey a small piece of information that explains why an 
                answer is correct or incorrect. When an answer with a feedback setting is selected and graded, a small box containing the feedback is displayed next to the answer.</p>
```

**Review by Cris-Moran — COMMENTED** [review]
Sorry, this was a mistake. Please dismiss this review.

File: elements/pl-multiple-choice/pl-multiple-choice.py
Comment (Cris-Moran) [inline elements/pl-multiple-choice/pl-multiple-choice.py:347]: One thing I'd like to bring up: in the `grade()` function, the `data['submitted_answers']` object doesn't contain the feedback I inserted into the answer tuples, so at the moment I'm not sure if there's another way to grab the feedback from the submitted answer without using a for loop.

Comment (jonatanschroeder) [reply elements/pl-multiple-choice/pl-multiple-choice.py:347]: I believe this is a reasonable solution.

**Comment by jonatanschroeder** [comment]
It seems tests are failing because you haven't updated the `test` function. You should update `data['partial_scores']` with the value that is expected to be set by `grade`, which includes the feedback.

**Comment by Cris-Moran** [comment]
Ah I see. I'll fix that shortly.

**Comment by Cris-Moran** [comment]
I fixed the issue by setting feedback to `None` in the `test()` function

**Comment by Cris-Moran** [comment]
Hmm it seems to still be failing. Looks like I need to put something other than `None` into the `data['partial_scores']` object

File: elements/pl-multiple-choice/pl-multiple-choice.py
Comment (jonatanschroeder) [inline elements/pl-multiple-choice/pl-multiple-choice.py:370]:
```suggestion
        feedback = data['correct_answers'][name].get('feedback', None)
        data['partial_scores'][name] = {'score': 1, 'weight': weight, 'feedback': feedback}
```

I'm not 100% sure this would work, but you should test it.

Comment (Cris-Moran) [reply elements/pl-multiple-choice/pl-multiple-choice.py:370]: I tested this locally and it appears to work. I'll go ahead and commit it.

**Review by jonatanschroeder — APPROVED** [review]

**Comment by Cris-Moran** [comment]
Before this is merged, I'd like to propose changing the spacing of the feedback again. I tested the feature with a different question and to me it looks cramped. For this reason I'd like to change the `margin-left` of the feedback from `0.5rem` (first image) to `0.75rem` (second image). It's a subtle difference, but I think it looks better.
Screen Shot 2021-08-18 at 6 03 15 PM
Screen Shot 2021-08-18 at 6 01 38 PM

File: elements/pl-multiple-choice/pl-multiple-choice.mustache
Comment (mwest1066) [inline elements/pl-multiple-choice/pl-multiple-choice.mustache:21]: I love the display of feedback in the question panel next to the selected item. But in the case where the student makes multiple submissions, the feedback will only be visible for the most recent submission (in the question panel). Can we additionally display the feedback in the submission panel, next to the submitted answer? That way the student will be able to see the history of their feedback together with their history of submitted answers.

File: docs/elements.md
Comment (mwest1066) [inline docs/elements.md:134]: The feedback is actually HTML, I think? That is, we template it in with triple-braces. Maybe clarify that here?
```suggestion
`feedback` | string | — | Helper text (HTML) to be displayed to the student next to the option after question is graded if this option has been selected by the student.
```

Comment (mwest1066) [reply docs/elements.md:134]: What happens if the feedback is very long? Does it look horrible? I wonder whether we should add an option like `show-feedback-at-end` to put the feedback after the multiple choice element? Maybe I'm overthinking this, however.

Comment (Cris-Moran) [reply docs/elements.md:134]: Here's what a long hint looks like:
Screen Shot 2021-08-30 at 8 43 49 PM
For an example of how this looks on a smaller screen, see https://github.com/PrairieLearn/PrairieLearn/pull/4656#issuecomment-898783046

File: elements/pl-multiple-choice/pl-multiple-choice.mustache
Comment (Cris-Moran) [reply elements/pl-multiple-choice/pl-multiple-choice.mustache:21]: Just implemented this, here's what it looks like:
Screen Shot 2021-08-30 at 8 49 01 PM

**Comment by Cris-Moran** [comment]
Just pushed all the requested changes.

File: docs/elements.md
Comment (mwest1066) [reply docs/elements.md]: Ok, that's kind of horrible but at least you can still read it 😄 --- I'm fine with this.

File: elements/pl-multiple-choice/pl-multiple-choice.mustache
Comment (mwest1066) [reply elements/pl-multiple-choice/pl-multiple-choice.mustache:21]: Perfect!

**Review by mwest1066 — APPROVED** [review]

**Comment by mwest1066** [comment]
@jonatanschroeder please merge when you are ready.

File: exampleCourse/questions/element/checkbox/question.html
Comment (jonatanschroeder) [inline exampleCourse/questions/element/checkbox/question.html:284]: @Cris-Moran did you add this to the wrong PR?

Comment (Cris-Moran) [reply exampleCourse/questions/element/checkbox/question.html:284]: I don't remember writing this. I believe this appeared when I pulled from upstream.

Comment (jonatanschroeder) [reply exampleCourse/questions/element/checkbox/question.html]: Doesn't matter, I removed it, it was causing tests to fail and is not relevant anyway.

**Pull request merged**
